File: Cpp/Session.cpp
#include "Session.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QHash>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QSet>
#include <QStringConverter>
#include <QTextStream>
#include <algorithm>

namespace Session {

static const int kMaxVersions = 10;

QString configDir()
{
    return QDir::homePath() + "/.noteeditor";
}

static QString historyDir()
{
    return configDir() + "/history";
}

static void saveHistories(const QVector<TabSnapshot> &tabs);

QString draftsDir()
{
    return configDir() + "/drafts";
}

QString docsDir()
{
    return configDir() + "/docs";
}

QString trashDir()
{
    return configDir() + "/trash";
}

QString versionsDir()
{
    return configDir() + "/versions";
}

static QString sessionFile()
{
    return configDir() + "/session.json";
}

static QString indexFile()
{
    return configDir() + "/index.json";
}

static QString trashIndexFile()
{
    return configDir() + "/trash_index.json";
}

static QString recentFile()
{
    return configDir() + "/recent.json";
}

static QString windowFile()
{
    return configDir() + "/window.json";
}

static QJsonValue toJsonOrNull(const QString &s)
{
    return s.isEmpty() ? QJsonValue(QJsonValue::Null) : QJsonValue(s);
}

static QString fromJsonOrEmpty(const QJsonValue &v)
{
    return v.isString() ? v.toString() : QString();
}

static QJsonObject loadJsonObject(const QString &path)
{
    QFile f(path);
    if (!f.open(QIODevice::ReadOnly))
        return QJsonObject();
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    return doc.isObject() ? doc.object() : QJsonObject();
}

static bool writeJsonObject(const QString &path, const QJsonObject &obj)
{
    QFile f(path);
    if (!f.open(QIODevice::WriteOnly | QIODevice::Truncate))
        return false;
    f.write(QJsonDocument(obj).toJson(QJsonDocument::Indented));
    return true;
}

static bool writeTextFile(const QString &path, const QString &content)
{
    QFile f(path);
    if (!f.open(QIODevice::WriteOnly | QIODevice::Truncate))
        return false;
    QTextStream out(&f);
    out.setEncoding(QStringConverter::Utf8);
    out << content;
    return true;
}

static QJsonObject metaToJson(const QString &filePath, const QString &defaultName, bool modified)
{
    QJsonObject o;
    o["file_path"] = toJsonOrNull(filePath);
    o["default_name"] = toJsonOrNull(defaultName);
    o["modified"] = modified;
    return o;
}

// The "pinned" flag is owned by setPinned(), not by the tab snapshots that
// rewrite the rest of an index entry: carry it over.
static QJsonObject mergedMeta(const QJsonObject &index, const TabSnapshot &info)
{
    QJsonObject meta = metaToJson(info.filePath, info.defaultName, info.modified);
    if (index.value(info.id).toObject().value("pinned").toBool(false))
        meta["pinned"] = true;
    if (index.value(info.id).toObject().value("remember").toBool(false))
        meta["remember"] = true;
    return meta;
}

void saveSession(const QVector<TabSnapshot> &tabs, const QString &activeId)
{
    QDir().mkpath(draftsDir());

    QJsonObject index = loadJsonObject(indexFile());
    QJsonArray entries;

    for (const TabSnapshot &info : tabs) {
        writeTextFile(draftsDir() + "/" + info.id + ".txt", info.content);

        QJsonObject entry = metaToJson(info.filePath, info.defaultName, info.modified);
        entry["id"] = info.id;
        if (info.cursor >= 0) {
            // où l'on travaillait dans la note (session.json seulement, pas index.json)
            entry["cursor"] = info.cursor;
            entry["scroll"] = info.scroll;
            if (info.anchor >= 0 && info.anchor != info.cursor)
                entry["anchor"] = info.anchor;
        }
        entries.append(entry);

        index[info.id] = mergedMeta(index, info);
    }

    QJsonObject sessionObj;
    sessionObj["active_id"] = toJsonOrNull(activeId);
    sessionObj["tabs"] = entries;
    writeJsonObject(sessionFile(), sessionObj);

    writeJsonObject(indexFile(), index);

    saveHistories(tabs);
}

// Un fichier `history/<id>.json` par onglet ouvert ayant un historique annuler/
// rétablir ; les autres (onglets fermés depuis, historique vide) sont supprimés.
static void saveHistories(const QVector<TabSnapshot> &tabs)
{
    QSet<QString> wanted;
    for (const TabSnapshot &info : tabs) {
        if (info.history.isEmpty())
            continue;
        wanted.insert(info.id);
        QDir().mkpath(historyDir());
        QFile f(historyDir() + "/" + info.id + ".json");
        if (f.open(QIODevice::WriteOnly | QIODevice::Truncate))
            f.write(QJsonDocument(info.history).toJson(QJsonDocument::Compact));
    }
    const QStringList names = QDir(historyDir()).entryList({"*.json"}, QDir::Files);
    for (const QString &name : names) {
        if (!wanted.contains(name.chopped(5)))
            QFile::remove(historyDir() + "/" + name);
    }
}

QJsonObject loadHistory(const QString &id)
{
    return loadJsonObject(historyDir() + "/" + id + ".json");
}

void saveDraft(const TabSnapshot &info)
{
    QDir().mkpath(draftsDir());
    writeTextFile(draftsDir() + "/" + info.id + ".txt", info.content);

    QJsonObject index = loadJsonObject(indexFile());
    index[info.id] = mergedMeta(index, info);
    writeJsonObject(indexFile(), index);
}

QVector<TabSnapshot> loadSession(QString *activeId)
{
    if (activeId)
        activeId->clear();

    QFile f(sessionFile());
    if (!f.open(QIODevice::ReadOnly))
        return {};
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    if (!doc.isObject())
        return {};
    const QJsonObject obj = doc.object();

    if (activeId)
        *activeId = fromJsonOrEmpty(obj.value("active_id"));

    QVector<TabSnapshot> result;
    for (const QJsonValue &v : obj.value("tabs").toArray()) {
        const QJsonObject entry = v.toObject();
        const QString id = entry.value("id").toString();

        QFile draftFile(draftsDir() + "/" + id + ".txt");
        if (!draftFile.open(QIODevice::ReadOnly))
            continue;
        const QString content = QString::fromUtf8(draftFile.readAll());

        TabSnapshot snap;
        snap.id = id;
        snap.filePath = fromJsonOrEmpty(entry.value("file_path"));
        snap.defaultName = fromJsonOrEmpty(entry.value("default_name"));
        snap.modified = entry.value("modified").toBool(false);
        snap.content = content;
        if (entry.contains("cursor")) {
            snap.cursor = entry.value("cursor").toInt(-1);
            snap.scroll = entry.value("scroll").toInt(0);
            if (entry.contains("anchor"))
                snap.anchor = entry.value("anchor").toInt(-1);
        }
        result.append(snap);
    }
    return result;
}

QVector<DraftEntry> listDrafts()
{
    QDir dir(draftsDir());
    if (!dir.exists())
        return {};

    const QJsonObject index = loadJsonObject(indexFile());

    QHash<QString, QJsonObject> sessionMeta;
    {
        QFile f(sessionFile());
        if (f.open(QIODevice::ReadOnly)) {
            const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
            if (doc.isObject()) {
                for (const QJsonValue &v : doc.object().value("tabs").toArray()) {
                    const QJsonObject entry = v.toObject();
                    sessionMeta[entry.value("id").toString()] = entry;
                }
            }
        }
    }

    QVector<DraftEntry> items;
    const QStringList files = dir.entryList(QStringList() << "*.txt", QDir::Files);
    for (const QString &name : files) {
        const QString id = name.left(name.length() - 4);
        const QFileInfo fi(dir.filePath(name));

        QJsonObject meta = index.value(id).toObject();
        if (meta.isEmpty() && sessionMeta.contains(id))
            meta = sessionMeta.value(id);

        DraftEntry entry;
        entry.id = id;
        entry.filePath = fromJsonOrEmpty(meta.value("file_path"));
        entry.defaultName = fromJsonOrEmpty(meta.value("default_name"));
        entry.modified = meta.contains("modified") ? meta.value("modified").toBool(true) : true;
        entry.pinned = meta.value("pinned").toBool(false);
        entry.remember = meta.value("remember").toBool(false);
        entry.mtimeMs = fi.lastModified().toMSecsSinceEpoch();
        items.append(entry);
    }

    std::sort(items.begin(), items.end(), [](const DraftEntry &a, const DraftEntry &b) {
        return a.mtimeMs > b.mtimeMs;
    });
    return items;
}

static QJsonArray loadRecentArray()
{
    return loadJsonObject(recentFile()).value("recent").toArray();
}

int recentLimit()
{
    const QJsonValue v = loadJsonObject(recentFile()).value("limit");
    if (v.isDouble() && v.toDouble() == static_cast<int>(v.toDouble()))
        return qBound(kMinRecent, v.toInt(), kMaxRecent);
    return kDefaultRecent;
}

static void saveRecentArray(const QJsonArray &entries, int limit = -1)
{
    QDir().mkpath(configDir());
    QJsonObject obj; // la limite est relue AVANT d'écrire (l'écriture remplace le fichier)
    obj["limit"] = limit < 0 ? recentLimit() : limit;
    obj["recent"] = entries;
    writeJsonObject(recentFile(), obj);
}

void setRecentLimit(int limit)
{
    limit = qBound(kMinRecent, limit, kMaxRecent);
    QJsonArray kept;
    for (const QJsonValue &v : loadRecentArray()) {
        if (kept.size() < limit)
            kept.append(v);
    }
    saveRecentArray(kept, limit);
}

void addRecent(const DraftEntry &entry)
{
    const int limit = recentLimit();
    QJsonArray entries;
    QJsonObject fresh;
    fresh["id"] = entry.id;
    fresh["file_path"] = toJsonOrNull(entry.filePath);
    fresh["default_name"] = toJsonOrNull(entry.defaultName);
    entries.append(fresh);
    for (const QJsonValue &v : loadRecentArray()) {
        if (v.toObject().value("id").toString() != entry.id && entries.size() < limit)
            entries.append(v);
    }
    saveRecentArray(entries);
}

void removeRecent(const QString &draftId)
{
    const QJsonArray entries = loadRecentArray();
    QJsonArray kept;
    for (const QJsonValue &v : entries) {
        if (v.toObject().value("id").toString() != draftId)
            kept.append(v);
    }
    if (kept.size() != entries.size())
        saveRecentArray(kept);
}

void clearRecent()
{
    if (!loadRecentArray().isEmpty())
        saveRecentArray(QJsonArray());
}

QVector<DraftEntry> listRecent()
{
    QSet<QString> drafts;
    for (const DraftEntry &d : listDrafts())
        drafts.insert(d.id);
    QVector<DraftEntry> result;
    for (const QJsonValue &v : loadRecentArray()) {
        const QJsonObject o = v.toObject();
        DraftEntry e;
        e.id = o.value("id").toString();
        if (e.id.isEmpty())
            continue;
        e.filePath = fromJsonOrEmpty(o.value("file_path"));
        e.defaultName = fromJsonOrEmpty(o.value("default_name"));
        const bool valid = (!e.filePath.isEmpty() && isExternalFile(e.filePath)) ? QFileInfo::exists(e.filePath)
                                                                                : drafts.contains(e.id);
        if (valid)
            result.append(e);
    }
    return result;
}

bool isExternalFile(const QString &filePath)
{
    if (filePath.isEmpty())
        return false;
    const QString path = QDir::cleanPath(QFileInfo(filePath).absoluteFilePath());
    const QString root = QDir::cleanPath(QFileInfo(configDir()).absoluteFilePath());
    return !(path == root || path.startsWith(root + "/"));
}

DraftEntry findDraftForPath(const QString &filePath)
{
    for (const DraftEntry &entry : listDrafts()) { // newest first
        if (entry.filePath == filePath)
            return entry;
    }
    return DraftEntry();
}

QString readDraft(const QString &draftId)
{
    QFile f(draftsDir() + "/" + draftId + ".txt");
    if (!f.open(QIODevice::ReadOnly))
        return QString();
    return QString::fromUtf8(f.readAll());
}

void deleteDraft(const QString &draftId)
{
    QFile::remove(draftsDir() + "/" + draftId + ".txt");

    QJsonObject index = loadJsonObject(indexFile());
    if (index.contains(draftId)) {
        index.remove(draftId);
        writeJsonObject(indexFile(), index);
    }
    removeRecent(draftId);
}

bool isPinned(const QString &draftId)
{
    return loadJsonObject(indexFile()).value(draftId).toObject().value("pinned").toBool(false);
}

void setPinned(const QString &draftId, bool pinned)
{
    QJsonObject index = loadJsonObject(indexFile());
    QJsonObject entry = index.value(draftId).toObject();
    if (pinned)
        entry["pinned"] = true;
    else
        entry.remove("pinned");
    index[draftId] = entry;
    writeJsonObject(indexFile(), index);
}

bool isRemembered(const QString &draftId)
{
    return loadJsonObject(indexFile()).value(draftId).toObject().value("remember").toBool(false);
}

void setRemembered(const QString &draftId, bool remembered)
{
    QJsonObject index = loadJsonObject(indexFile());
    QJsonObject entry = index.value(draftId).toObject();
    if (remembered)
        entry["remember"] = true;
    else
        entry.remove("remember");
    index[draftId] = entry;
    writeJsonObject(indexFile(), index);
}

void renameDraft(const QString &draftId, const QString &newDefaultName)
{
    QJsonObject index = loadJsonObject(indexFile());
    QJsonObject entry = index.value(draftId).toObject();
    entry["default_name"] = newDefaultName;
    index[draftId] = entry;
    writeJsonObject(indexFile(), index);
}

void trashDraft(const QString &draftId)
{
    QDir().mkpath(trashDir());
    const QString src = draftsDir() + "/" + draftId + ".txt";
    const QString dst = trashDir() + "/" + draftId + ".txt";
    QFile::remove(dst);
    if (!QFile::rename(src, dst))
        return;

    QJsonObject index = loadJsonObject(indexFile());
    QJsonObject meta = index.value(draftId).toObject();
    index.remove(draftId);
    writeJsonObject(indexFile(), index);
    removeRecent(draftId);

    meta["deleted_at"] = QDateTime::currentDateTime().toString(Qt::ISODate);
    QJsonObject trashIndex = loadJsonObject(trashIndexFile());
    trashIndex[draftId] = meta;
    writeJsonObject(trashIndexFile(), trashIndex);
}

QVector<TrashEntry> listTrash()
{
    const QJsonObject trashIndex = loadJsonObject(trashIndexFile());
    QVector<TrashEntry> items;
    for (auto it = trashIndex.constBegin(); it != trashIndex.constEnd(); ++it) {
        const QJsonObject meta = it.value().toObject();
        TrashEntry entry;
        entry.id = it.key();
        entry.filePath = fromJsonOrEmpty(meta.value("file_path"));
        entry.defaultName = fromJsonOrEmpty(meta.value("default_name"));
        entry.deletedAt = fromJsonOrEmpty(meta.value("deleted_at"));
        items.append(entry);
    }
    std::sort(items.begin(), items.end(), [](const TrashEntry &a, const TrashEntry &b) {
        return a.deletedAt > b.deletedAt;
    });
    return items;
}

void restoreDraft(const QString &draftId)
{
    const QString src = trashDir() + "/" + draftId + ".txt";
    const QString dst = draftsDir() + "/" + draftId + ".txt";
    QDir().mkpath(draftsDir());
    QFile::remove(dst);
    if (!QFile::rename(src, dst))
        return;

    QJsonObject trashIndex = loadJsonObject(trashIndexFile());
    QJsonObject meta = trashIndex.value(draftId).toObject();
    trashIndex.remove(draftId);
    writeJsonObject(trashIndexFile(), trashIndex);
    meta.remove("deleted_at");

    QJsonObject index = loadJsonObject(indexFile());
    index[draftId] = meta;
    writeJsonObject(indexFile(), index);
}

void purgeDraft(const QString &draftId)
{
    QFile::remove(trashDir() + "/" + draftId + ".txt");
    QJsonObject trashIndex = loadJsonObject(trashIndexFile());
    if (trashIndex.contains(draftId)) {
        trashIndex.remove(draftId);
        writeJsonObject(trashIndexFile(), trashIndex);
    }
}

static void pruneVersions(const QString &draftId)
{
    const QString dir = versionsDir() + "/" + draftId;
    QStringList names = QDir(dir).entryList(QStringList() << "*.txt", QDir::Files, QDir::Name | QDir::Reversed);
    for (int i = kMaxVersions; i < names.size(); ++i)
        QFile::remove(dir + "/" + names[i]);
}

void saveVersion(const QString &draftId, const QString &content)
{
    const QString dir = versionsDir() + "/" + draftId;
    QDir().mkpath(dir);
    const QString stamp = QDateTime::currentDateTime().toString("yyyyMMddTHHmmsszzz");
    writeTextFile(dir + "/" + stamp + ".txt", content);
    pruneVersions(draftId);
}

void mergeVersions(const QString &fromId, const QString &toId)
{
    const QString src = versionsDir() + "/" + fromId;
    if (!QDir(src).exists())
        return;
    const QString dst = versionsDir() + "/" + toId;
    QDir().mkpath(dst);
    for (const QString &name : QDir(src).entryList(QDir::Files)) {
        QFile::remove(dst + "/" + name);
        QFile::rename(src + "/" + name, dst + "/" + name);
    }
    QDir().rmdir(src);
    pruneVersions(toId);
}

QStringList listVersions(const QString &draftId)
{
    const QString dir = versionsDir() + "/" + draftId;
    if (!QDir(dir).exists())
        return {};
    QStringList names = QDir(dir).entryList(QStringList() << "*.txt", QDir::Files, QDir::Name | QDir::Reversed);
    QStringList stamps;
    for (const QString &name : names)
        stamps.append(name.left(name.length() - 4));
    return stamps;
}

QString readVersion(const QString &draftId, const QString &stamp)
{
    QFile f(versionsDir() + "/" + draftId + "/" + stamp + ".txt");
    if (!f.open(QIODevice::ReadOnly))
        return QString();
    return QString::fromUtf8(f.readAll());
}

void saveWindowState(int width, int height, const QList<int> &splitterSizes, int x, int y, bool wordWrap,
                     bool markdownPreview, bool sidebarVisible)
{
    QDir().mkpath(configDir());
    QJsonObject obj;
    obj["width"] = width;
    obj["height"] = height;
    QJsonArray sizesArray;
    for (int s : splitterSizes)
        sizesArray.append(s);
    obj["splitter_sizes"] = sizesArray;
    obj["x"] = x;
    obj["y"] = y;
    obj["word_wrap"] = wordWrap;
    obj["markdown_preview"] = markdownPreview;
    obj["sidebar_visible"] = sidebarVisible;
    writeJsonObject(windowFile(), obj);
}

WindowState loadWindowState()
{
    WindowState state;
    QFile f(windowFile());
    if (!f.open(QIODevice::ReadOnly))
        return state;
    const QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    if (!doc.isObject())
        return state;

    const QJsonObject obj = doc.object();
    state.width = obj.value("width").toInt(0);
    state.height = obj.value("height").toInt(0);
    for (const QJsonValue &v : obj.value("splitter_sizes").toArray())
        state.splitterSizes.append(v.toInt());
    state.hasPosition = obj.contains("x") && obj.contains("y");
    if (state.hasPosition) {
        state.x = obj.value("x").toInt(0);
        state.y = obj.value("y").toInt(0);
    }
    state.hasSidebar = obj.contains("sidebar_visible");
    state.sidebarVisible = obj.value("sidebar_visible").toBool(true);
    state.hasWordWrap = obj.contains("word_wrap");
    state.wordWrap = obj.value("word_wrap").toBool(true);
    state.hasMarkdownPreview = obj.contains("markdown_preview");
    state.markdownPreview = obj.value("markdown_preview").toBool(false);
    state.valid = state.width > 0 && state.height > 0;
    return state;
}

} // namespace Session

File: Python/session.py
import datetime
import json
import os

CONFIG_DIR = os.path.expanduser("~/.noteeditor")
DRAFTS_DIR = os.path.join(CONFIG_DIR, "drafts")
DOCS_DIR = os.path.join(CONFIG_DIR, "docs")
TRASH_DIR = os.path.join(CONFIG_DIR, "trash")
VERSIONS_DIR = os.path.join(CONFIG_DIR, "versions")
SESSION_FILE = os.path.join(CONFIG_DIR, "session.json")
INDEX_FILE = os.path.join(CONFIG_DIR, "index.json")
TRASH_INDEX_FILE = os.path.join(CONFIG_DIR, "trash_index.json")
WINDOW_FILE = os.path.join(CONFIG_DIR, "window.json")
RECENT_FILE = os.path.join(CONFIG_DIR, "recent.json")

MAX_VERSIONS = 10
DEFAULT_RECENT = 10
MIN_RECENT = 1
MAX_RECENT = 50


def _load_json(path, default):
    if not os.path.isfile(path):
        return default
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (OSError, ValueError):
        return default


def _load_index():
    return _load_json(INDEX_FILE, {})


def _save_index(index):
    with open(INDEX_FILE, "w", encoding="utf-8") as f:
        json.dump(index, f, ensure_ascii=False, indent=2)


def _load_trash_index():
    return _load_json(TRASH_INDEX_FILE, {})


def _save_trash_index(index):
    with open(TRASH_INDEX_FILE, "w", encoding="utf-8") as f:
        json.dump(index, f, ensure_ascii=False, indent=2)


def _merged_meta(index, info):
    """Index entry for a tab, keeping the "pinned" flag it may already carry:
    it is owned by set_pinned(), not by the tab snapshots that rewrite the rest."""
    meta = {
        "file_path": info["file_path"],
        "default_name": info["default_name"],
        "modified": info["modified"],
    }
    if index.get(info["id"], {}).get("pinned"):
        meta["pinned"] = True
    if index.get(info["id"], {}).get("remember"):
        meta["remember"] = True
    return meta


def is_pinned(draft_id):
    return bool(_load_index().get(draft_id, {}).get("pinned"))


def set_pinned(draft_id, pinned):
    """Pins/unpins a draft. A pinned draft can be neither closed nor trashed."""
    index = _load_index()
    entry = index.get(draft_id, {})
    if pinned:
        entry["pinned"] = True
    else:
        entry.pop("pinned", None)
    index[draft_id] = entry
    _save_index(index)


def is_remembered(draft_id):
    """True when the note asked to be remembered in the recently-closed list even
    if it is empty (an empty untitled note is skipped otherwise)."""
    return bool(_load_index().get(draft_id, {}).get("remember"))


def set_remembered(draft_id, remembered):
    index = _load_index()
    entry = index.get(draft_id, {})
    if remembered:
        entry["remember"] = True
    else:
        entry.pop("remember", None)
    index[draft_id] = entry
    _save_index(index)


def save_session(tabs_info, active_id=None):
    os.makedirs(DRAFTS_DIR, exist_ok=True)

    entries = []
    index = _load_index()
    for info in tabs_info:
        draft_path = os.path.join(DRAFTS_DIR, info["id"] + ".txt")
        with open(draft_path, "w", encoding="utf-8") as f:
            f.write(info["content"])
        entry = {
            "id": info["id"],
            "file_path": info["file_path"],
            "default_name": info["default_name"],
            "modified": info["modified"],
        }
        if "cursor" in info:
            # où l'on travaillait dans la note (session.json seulement, pas index.json)
            entry["cursor"] = info["cursor"]
            entry["scroll"] = info.get("scroll", 0)
        entries.append(entry)
        index[info["id"]] = _merged_meta(index, info)

    with open(SESSION_FILE, "w", encoding="utf-8") as f:
        json.dump({"active_id": active_id, "tabs": entries}, f, ensure_ascii=False, indent=2)

    _save_index(index)


def save_draft(info):
    """Persist a single tab's content to the drafts store, without touching the active session list."""
    os.makedirs(DRAFTS_DIR, exist_ok=True)
    draft_path = os.path.join(DRAFTS_DIR, info["id"] + ".txt")
    with open(draft_path, "w", encoding="utf-8") as f:
        f.write(info["content"])

    index = _load_index()
    index[info["id"]] = _merged_meta(index, info)
    _save_index(index)


def rename_draft(draft_id, new_default_name):
    """Renames an untitled draft (one with no real file_path)."""
    index = _load_index()
    entry = index.get(draft_id, {})
    entry["default_name"] = new_default_name
    index[draft_id] = entry
    _save_index(index)


def load_session():
    data = _load_json(SESSION_FILE, None)
    if data is None:
        return [], None

    if isinstance(data, list):
        data = {"active_id": None, "tabs": data}

    result = []
    for entry in data.get("tabs", []):
        draft_path = os.path.join(DRAFTS_DIR, entry["id"] + ".txt")
        try:
            with open(draft_path, "r", encoding="utf-8") as f:
                content = f.read()
        except OSError:
            continue
        result.append({**entry, "content": content})
    return result, data.get("active_id")


def list_drafts():
    """All draft files on disk, newest first, with the best label metadata available."""
    if not os.path.isdir(DRAFTS_DIR):
        return []

    index = _load_index()
    session_data = _load_json(SESSION_FILE, {})
    if isinstance(session_data, list):
        session_data = {"tabs": session_data}
    session_meta = {entry["id"]: entry for entry in session_data.get("tabs", [])}

    items = []
    for name in os.listdir(DRAFTS_DIR):
        if not name.endswith(".txt"):
            continue
        draft_id = name[:-4]
        path = os.path.join(DRAFTS_DIR, name)
        try:
            mtime = os.path.getmtime(path)
        except OSError:
            continue
        meta = index.get(draft_id) or session_meta.get(draft_id) or {}
        items.append(
            {
                "id": draft_id,
                "file_path": meta.get("file_path"),
                "default_name": meta.get("default_name"),
                "modified": meta.get("modified", True),
                "pinned": bool(meta.get("pinned")),
                "remember": bool(meta.get("remember")),
                "mtime": mtime,
            }
        )

    items.sort(key=lambda e: e["mtime"], reverse=True)
    return items


def is_external_file(file_path):
    """True for a real file living outside ~/.noteeditor. Its text is safe in the
    file itself, so once closed it no longer needs a spot in the drafts sidebar;
    untitled notes and files under ~/.noteeditor (docs/) keep theirs."""
    if not file_path:
        return False
    path = os.path.normpath(os.path.abspath(file_path))
    root = os.path.normpath(os.path.abspath(CONFIG_DIR))
    return not (path == root or path.startswith(root + os.sep))


def _load_recent():
    data = _load_json(RECENT_FILE, {})
    entries = data.get("recent", []) if isinstance(data, dict) else []
    return [e for e in entries if isinstance(e, dict) and e.get("id")]


def get_recent_limit():
    """How many recently closed notes are remembered (recent.json "limit")."""
    data = _load_json(RECENT_FILE, {})
    limit = data.get("limit") if isinstance(data, dict) else None
    if isinstance(limit, int) and not isinstance(limit, bool):
        return max(MIN_RECENT, min(MAX_RECENT, limit))
    return DEFAULT_RECENT


def _save_recent(entries, limit=None):
    os.makedirs(CONFIG_DIR, exist_ok=True)
    # relire la limite AVANT d'ouvrir le fichier en écriture (qui le vide)
    data = {"recent": entries, "limit": get_recent_limit() if limit is None else limit}
    with open(RECENT_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


def set_recent_limit(limit):
    """Changes the size of the recently-closed list; the list is trimmed at once."""
    limit = max(MIN_RECENT, min(MAX_RECENT, int(limit)))
    _save_recent(_load_recent()[:limit], limit)


def add_recent(entry):
    """Remembers a note that was just closed (newest first, at most get_recent_limit()).
    entry: {"id", "file_path", "default_name"}."""
    entries = [e for e in _load_recent() if e["id"] != entry["id"]]
    entries.insert(
        0,
        {"id": entry["id"], "file_path": entry.get("file_path"), "default_name": entry.get("default_name")},
    )
    _save_recent(entries[: get_recent_limit()])


def remove_recent(draft_id):
    entries = _load_recent()
    kept = [e for e in entries if e["id"] != draft_id]
    if len(kept) != len(entries):
        _save_recent(kept)


def clear_recent():
    if _load_recent():
        _save_recent([])


def list_recent():
    """Recently closed notes that can still be reopened: a file outside
    ~/.noteeditor must still exist, any other note must still have its draft."""
    drafts = {e["id"] for e in list_drafts()}
    result = []
    for entry in _load_recent():
        path = entry.get("file_path")
        if path and is_external_file(path):
            valid = os.path.isfile(path)
        else:
            valid = entry["id"] in drafts
        if valid:
            result.append(entry)
    return result


def find_draft_for_path(file_path):
    """The most recent draft bound to this file, or None. Opening a file must
    reuse it rather than mint a new draft, or every open/close cycle of the same
    file adds another entry to the drafts sidebar."""
    for entry in list_drafts():  # newest first
        if entry["file_path"] == file_path:
            return entry
    return None


def read_draft(draft_id):
    path = os.path.join(DRAFTS_DIR, draft_id + ".txt")
    with open(path, "r", encoding="utf-8") as f:
        return f.read()


def delete_draft(draft_id):
    """Permanently removes a draft (file + index entry), no trash involved."""
    path = os.path.join(DRAFTS_DIR, draft_id + ".txt")
    try:
        os.remove(path)
    except OSError:
        pass
    index = _load_index()
    if index.pop(draft_id, None) is not None:
        _save_index(index)
    remove_recent(draft_id)


def trash_draft(draft_id):
    """Moves a draft to the trash instead of deleting it outright."""
    os.makedirs(TRASH_DIR, exist_ok=True)
    src = os.path.join(DRAFTS_DIR, draft_id + ".txt")
    dst = os.path.join(TRASH_DIR, draft_id + ".txt")
    try:
        os.replace(src, dst)
    except OSError:
        return

    index = _load_index()
    meta = index.pop(draft_id, {})
    _save_index(index)
    remove_recent(draft_id)

    meta["deleted_at"] = datetime.datetime.now().isoformat()
    trash_index = _load_trash_index()
    trash_index[draft_id] = meta
    _save_trash_index(trash_index)


def list_trash():
    """All trashed drafts, most recently deleted first."""
    trash_index = _load_trash_index()
    items = []
    for draft_id, meta in trash_index.items():
        items.append(
            {
                "id": draft_id,
                "file_path": meta.get("file_path"),
                "default_name": meta.get("default_name"),
                "deleted_at": meta.get("deleted_at", ""),
            }
        )
    items.sort(key=lambda e: e["deleted_at"], reverse=True)
    return items


def restore_draft(draft_id):
    """Moves a trashed draft back into the active drafts store."""
    src = os.path.join(TRASH_DIR, draft_id + ".txt")
    dst = os.path.join(DRAFTS_DIR, draft_id + ".txt")
    os.makedirs(DRAFTS_DIR, exist_ok=True)
    try:
        os.replace(src, dst)
    except OSError:
        return

    trash_index = _load_trash_index()
    meta = trash_index.pop(draft_id, {})
    _save_trash_index(trash_index)
    meta.pop("deleted_at", None)

    index = _load_index()
    index[draft_id] = meta
    _save_index(index)


def purge_draft(draft_id):
    """Permanently deletes a trashed draft."""
    path = os.path.join(TRASH_DIR, draft_id + ".txt")
    try:
        os.remove(path)
    except OSError:
        pass
    trash_index = _load_trash_index()
    if trash_index.pop(draft_id, None) is not None:
        _save_trash_index(trash_index)


def save_version(draft_id, content):
    """Snapshots a tab's pre-save content, keeping only the most recent MAX_VERSIONS."""
    version_dir = os.path.join(VERSIONS_DIR, draft_id)
    os.makedirs(version_dir, exist_ok=True)
    stamp = datetime.datetime.now().strftime("%Y%m%dT%H%M%S%f")
    with open(os.path.join(version_dir, stamp + ".txt"), "w", encoding="utf-8") as f:
        f.write(content)

    _prune_versions(draft_id)


def _prune_versions(draft_id):
    version_dir = os.path.join(VERSIONS_DIR, draft_id)
    try:
        versions = sorted(os.listdir(version_dir), reverse=True)
    except OSError:
        return
    for stale in versions[MAX_VERSIONS:]:
        try:
            os.remove(os.path.join(version_dir, stale))
        except OSError:
            pass


def merge_versions(from_id, to_id):
    """Moves the saved versions of one draft under another id (the newest
    MAX_VERSIONS are kept) — used when a note adopts the id of an existing note
    for the same file."""
    src = os.path.join(VERSIONS_DIR, from_id)
    if not os.path.isdir(src):
        return
    dst = os.path.join(VERSIONS_DIR, to_id)
    os.makedirs(dst, exist_ok=True)
    for name in os.listdir(src):
        try:
            os.replace(os.path.join(src, name), os.path.join(dst, name))
        except OSError:
            pass
    try:
        os.rmdir(src)
    except OSError:
        pass
    _prune_versions(to_id)


def list_versions(draft_id):
    """Timestamps (newest first) of the saved versions kept for a tab."""
    version_dir = os.path.join(VERSIONS_DIR, draft_id)
    if not os.path.isdir(version_dir):
        return []
    names = sorted((n for n in os.listdir(version_dir) if n.endswith(".txt")), reverse=True)
    return [n[:-4] for n in names]


def read_version(draft_id, stamp):
    path = os.path.join(VERSIONS_DIR, draft_id, stamp + ".txt")
    with open(path, "r", encoding="utf-8") as f:
        return f.read()


def save_window_state(width, height, splitter_sizes, x=None, y=None, word_wrap=None, markdown_preview=None):
    """Persists the window size, screen position, sidebar-splitter position and
    display settings (word wrap, Markdown preview) across launches. Everything
    but the size and splitter is optional: omitted from the file when unknown."""
    os.makedirs(CONFIG_DIR, exist_ok=True)
    state = {"width": width, "height": height, "splitter_sizes": list(splitter_sizes)}
    if x is not None and y is not None:
        state["x"] = x
        state["y"] = y
    if word_wrap is not None:
        state["word_wrap"] = bool(word_wrap)
    if markdown_preview is not None:
        state["markdown_preview"] = bool(markdown_preview)
    with open(WINDOW_FILE, "w", encoding="utf-8") as f:
        json.dump(state, f, ensure_ascii=False, indent=2)


def load_window_state():
    """Returns {"width", "height", "splitter_sizes"[, "x", "y", "word_wrap", "markdown_preview"]}
    or None if never saved."""
    return _load_json(WINDOW_FILE, None)
